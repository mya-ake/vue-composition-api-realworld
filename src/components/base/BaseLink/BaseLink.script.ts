import Vue from 'vue';
import { BaseLinkType } from './BaseLink.type';
import { isOutSideLink } from '@/utils/judgers';

type Tag = 'a' | 'router-link';

const selectTag = (to: string): Tag => {
  return isOutSideLink(to) ? 'a' : 'router-link';
};

const buildAttrsATag = (href: string) => {
  return {
    href,
    rel: 'noopener',
    target: '_blank',
  };
};

const buildAttrs = (
  tag: Tag,
  props: BaseLinkType.Props,
): { [key: string]: string } => {
  switch (tag) {
    case 'a':
      return buildAttrsATag(props.to);
    default:
      return {};
  }
};

export default Vue.extend({
  props: {
    to: {
      type: String,
      required: true,
    },

    activeClass: {
      type: String,
      default: undefined,
    },
  },

  render(h) {
    const tag = selectTag(this.to);
    const attrs = buildAttrs(tag, this.$props as BaseLinkType.Props);
    return h(tag, { props: this.$props, attrs }, this.$slots.default);
  },
});
