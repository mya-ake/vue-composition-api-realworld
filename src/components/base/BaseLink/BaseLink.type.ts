export namespace BaseLinkType {
  export interface Props {
    to: string;
    activeClass?: string;
  }
}
