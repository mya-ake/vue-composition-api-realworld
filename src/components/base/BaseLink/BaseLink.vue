<script lang="ts" src="./BaseLink.script.ts"></script>
