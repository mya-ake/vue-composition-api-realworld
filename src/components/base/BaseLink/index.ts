export { default as BaseLink } from './BaseLink.vue';
export * from './BaseLink.type';
