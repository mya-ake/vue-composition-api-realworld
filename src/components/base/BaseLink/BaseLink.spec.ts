import { shallowMount, Wrapper } from '@vue/test-utils';
import BaseLink from './BaseLink.vue';

describe('components/contents/BaseLink', () => {
  describe('mountable', () => {
    let wrapper: Wrapper<Vue>;

    beforeEach(() => {
      wrapper = shallowMount(BaseLink, {
        propsData: {
          to: '/',
        },
        stubs: {
          RouterLink: true,
        },
      });
    });

    it('is vue instance', () => {
      expect(wrapper.isVueInstance()).toBe(true);
    });

    it('snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('toによってタグの種類が変わる', () => {
    it('外部リンクのときはaタグ', () => {
      const wrapper = shallowMount(BaseLink, {
        propsData: {
          to: 'https://example.com',
        },
      });
      expect(wrapper.element.tagName).toBe('A');
    });

    it('内部リンクのときはrouter-link', () => {
      const wrapper = shallowMount(BaseLink, {
        propsData: {
          to: '/',
        },
        stubs: {
          RouterLink: true,
        },
      });
      expect(wrapper.element.tagName).toBe('ROUTERLINK-STUB');
    });
  });
});
