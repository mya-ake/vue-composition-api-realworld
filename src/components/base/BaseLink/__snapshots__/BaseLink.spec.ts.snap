// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/base/BaseLink mountable snapshot 1`] = `<routerlink-stub />`;
